components/server/src/websocket/websocket-connection-manager.ts

```diff
@@ -106,39 +164,62 @@ export class WebsocketConnectionManager<C extends GitpodClient, S extends Gitpod
         });
     }
 
-    protected createRateLimiter(request?: object): RateLimiter {
-        const expressReq = request as express.Request;
-        const user = expressReq.user as User;
+    protected getOrCreateClientContext(expressReq: express.Request): WebsocketClientContext<C, S> {
+        const { clientId, authLevel } = this.getClientId(expressReq);
+        let ctx = this.contexts.get(clientId);
+        if (!ctx) {
+            ctx = new WebsocketClientContext(clientId, authLevel);
+            this.contexts.set(clientId, ctx);
+            this.events.emit(EVENT_CLIENT_CONTEXT_CREATED, ctx);
+        }
+        return ctx;
+    }
+
+    protected getClientId(expressReq: express.Request): { clientId: string, authLevel: WebsocketAuthenticationLevel } {
```

Comment:
it would be good to understand how this is connected with the usage side of things. e.g. when is `anonymous` a valid `clientId`, maybe for not logged in users?  and what's the difference to a session-only scenario?

Reply:
First and foremost: We already have theses levels implicitly in `GitpodServerImpl` either `this.user` / `this.client` might be undefined. I did not introduce those, but just surfaced them.
I basically took over the `getClientId` method from how the ratelimiter tells apart clients, which is a nice approach.